### DISCUSSION TITLE
computeQpOffDiagJacobian when used with the tensor mechanics Master action

### DISCUSSION THREAD

**Discussion by kengwit** [idaholab/moose#26230] [Q&A Modules: Solid mechanics] [answered]
Hello,

I am trying to add off-diagonal terms corresponding to the off-diagonal block or stiffness matrix $K_{\mathbf{u},\epsilon^p}$, where $\mathbf{u}$=nodal displacement vector, $\epsilon^p$ = a coupled variable.

I am basing the implementation off the PhaseFieldFractureMechanicsOffDiag::computeQpOffDiagJacobian function in the kernel [PhaseFieldFractureMechanicsOffDiag.C](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/kernels/PhaseFieldFractureMechanicsOffDiag.C). My kernel is identical to the code of this function:
```
  if (_c_coupled && jvar == _c_var)
  {
    Real val = 0.0;
    for (unsigned int k = 0; k < 3; ++k)
      val += _d_stress_dc[_qp](_component, k) * _grad_test[_i][_qp](k);
    return val * _phi[_j][_qp];
  }
```
except that, in my case, ```_d_stress_dc``` would be replaced by ```_d_stress_depsP```. As you can see in this code, ```_component``` is a user input parameter. I am unsure of how I will be using this when the tensor mechanics Master action, which already defines the needed stress divergence kernels, is used. For e.g., in the 2D case, with the displacement variables being $u_x$ and $u_y$, how would Moose know that it would be assembling $K_{u_x^{\_i},\epsilon^{p,\_j}}$ when ```_component=0``` and $K_{u_y^{\_i},\epsilon^{p,\_j}}$ when ```_component=1```? 

Thanks
Kengwit

**Comment by GiudGiud** [comment]
Hello

So _component is meant to differentiate between x and y here right?
The kernel would simply be added twice, once with component=0 and once with component=1

Guillaume

**Reply by kengwit** [reply]
I understand it needs to be added twice. But how does Moose know the location of the elements in the matrix based on the user parameter _component?

**Reply by GiudGiud** [reply]
MOOSE knows it from the `variable` parameter most likely. There's two variables, once for each component right?

**Reply by kengwit** [reply]
Do you mean ```jvar```? I believe that jvar corresponds to the coupled variable $\epsilon^p$ (or $c$ for the phase field problem), not the displacement components. I checked that ```jvar``` remains constant at the value 2.

**Reply by GiudGiud** [reply]
I meant the variable parameter of the kernel. It's not that then

If the variable is eps_p, then I think the `_c_var` is the displacement? As you can see this extra term is only added when `jvar = _c_var` (in the if condition)

**Reply by GiudGiud** [reply]
@sveerara

**Reply by kengwit** [reply, marked as answer]
_c_var is the coupled variable. I think you mean _var in the base class? I think your explanation makes sense. Thank you.
